experimental/offline_attestation/README.md

```diff
@@ -0,0 +1,111 @@
+# Offline Attestation
+
+The current Oak Functions Runtime implementation uses an interactive handshake
+for setting up an attested end-to-end encrypted channel between the client and
+the runtime. For some scenarios it might be preferable to avoid the handshake.
+
+The offline attestation approach splits the attestation process into two
+distinct parts:
+
+- Key generation and export: A single asymmetric key-pair is generated. The
+  public key is bound to an attestation report that provides evidence about the
+  enclave that generated the private key. The public key and attestation report
+  are then made availble to clients.
+- Encrypted communications: Clients use the public key (once they are satisfied
+  by the evidence that only appropriate enclaves would have access to the
+  private key) to encrypt requests to the server. Clients also provide
+  additional information (e.g. the client's public key) in the request to allow
+  the server to encrypt the response.
+
+- This approach has some advantages compared to the interactive handshake-based
+  approach:
+  - Reduced bandwith and latency
+  - No need for sticky sessions in a load-balanced environment
+- And some disadvantages:
+  - No forward secrecy since the same public key is used for encrypting all
+    requests
+  - No built-in replay protection
```

Comment:
If we will send an encrypted pre-shared secret (encrypted with server's public key) as random value used in the shared key generation - can it help with replay protection?

I.e. the key will be generated as a `PRF(server_pub_key, random_pre_shared_secret)`

Reply:
Any type of random value (e.g. a nonce or key) could help, but the server must then keep track of all the nonces it has received to check whether it has seen the current one before. In the interactive handshake, each handshake is based on a different ephemeral key from the server. That, combined with the transcript and the state-machine nature of the handshaker means that reply will be detected at the protocol level.